src/wallet/wallet.cpp

```diff
@@ -1812,31 +1826,41 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate, b
-                const auto& tx = block.vtx[posInBlock];
-                if (AddToWalletIfInvolvingMe(tx, pindex->GetBlockHash(), posInBlock, fUpdate)) {
-                    myTxHashes.push_back(tx->GetHash());
-                    ret++;
+
+            {
+                LOCK2(cs_main, cs_wallet);
+                if (pindex && !chainActive.Contains(pindex)) {
+                    // Abort scan if current block is no longer active, to prevent
+                    // marking transactions as coming from the wrong block.
+                    break;
+                }
+                for (int posInBlock = 0; posInBlock < (int) block.vtx.size(); posInBlock++) {
+                    const auto& tx = block.vtx[posInBlock];
+                    if (AddToWalletIfInvolvingMe(tx, pindex->GetBlockHash(), posInBlock, fUpdate)) {
+                        myTxHashes.push_back(tx->GetHash());
+                        ret++;
+                    }
                 }
-            }
 
-            // Sapling
-            // This should never fail: we should always be able to get the tree
-            // state on the path to the tip of our chain
-            if (pindex->pprev) {
-                if (Params().GetConsensus().NetworkUpgradeActive(pindex->pprev->nHeight,  Consensus::UPGRADE_V5_0)) {
-                    SaplingMerkleTree saplingTree;
-                    assert(pcoinsTip->GetSaplingAnchorAt(pindex->pprev->hashFinalSaplingRoot, saplingTree));
-                    // Increment note witness caches
-                    ChainTipAdded(pindex, &block, saplingTree);
+                // Sapling
+                // This should never fail: we should always be able to get the tree
+                // state on the path to the tip of our chain
+                if (pindex->pprev) {
+                    if (Params().GetConsensus().NetworkUpgradeActive(pindex->pprev->nHeight, Consensus::UPGRADE_V5_0)) {
+                        SaplingMerkleTree saplingTree;
+                        assert(pcoinsTip->GetSaplingAnchorAt(pindex->pprev->hashFinalSaplingRoot, saplingTree));
+                        // Increment note witness caches
+                        ChainTipAdded(pindex, &block, saplingTree);
+                    }
                 }
-            }
 
-            pindex = chainActive.Next(pindex);
-            if (GetTime() >= nNow + 60) {
-                nNow = GetTime();
-                LogPrintf("Still rescanning. At block %d. Progress=%f\n", pindex->nHeight, Checkpoints::GuessVerificationProgress(pindex));
+                pindex = chainActive.Next(pindex);
+                gvp = Checkpoints::GuessVerificationProgress(pindex, false);
+                if (tip != chainActive.Tip()) {
+                    tip = chainActive.Tip();
+                    // in case the tip has changed, update progress max
+                    dProgressTip = Checkpoints::GuessVerificationProgress(tip, false);
+                }
```

Comment:
These updates are not needed when `pindex` becomes null (after the last iteration, before exiting the loop).

Imo, better to update `gvp` at the very beginning of the loop, and update `tip` and `dProgressTip` (if needed) right after `LOCK2(cs_main, cs_wallet)` (and you can be sure that pindex is not null there).